### ISSUE TITLE
Bot Request: Version Bumper 9000

### ISSUE THREAD

**Issue by hobinjk** [new-xkit/XKit#955]
Commits would no longer need version number bumps, eliminating the primary cause of PR invalidation. Instead, contributors would specify "This is a (patch|minor|major) change." as part of their commit message. If there are multiple files involved, the contributor would instead specify the filename, e.g. "This is a patch change to Extensions/xkit_patches.js."

This bot would replace Homu in the review flow, making it instead look like this:
> me: LGTM @VersionBumper9000 r+
> VersionBumper9000: @homu r+

VB9k then waits for homu to finish then adds the version bumping commits directly to the merged branch.

Note that VB9k would error on unspecified version number changes, need to play nice with manual version number bumps, and allow not bumping versions on files that don't have version numbers.

**Comment by nightpool** [comment]
Hmm. Can we fork homu instead to do this? don't like the idea of chaining
bots like this....

**Comment by 0xazure** [comment]
I don't think we'd have any option other than to fork Homu and inject additional behaviour into its process because of the Homu flow and how we have auto-building currently set up.

My understanding of the current process (correct me if I'm wrong):

- Someone @mentions Homu and `r+`s the PR
- Homu merges the PR into `auto` and runs the tests
- If the tests pass, Homu merges the PR into `master`
- Successful tests of the `master` branch (however they get triggered, usually by a Homu merge) automatically run our build steps [as a post-build script](https://github.com/new-xkit/XKit/blob/38def42023ac7de2335add89ac83f4171faeca72/.travis.yml#L7-L8)

Given this process, I don't see anywhere we could easily interrupt any of the steps for another tool to do any additional work on the files in a PR until well after a new distribution has already been built.  

Homu _is_ available at [barosl/homu](https://github.com/barosl/homu), we would just need to stand up a compatible python environment accessible at a public URL so GitHub can use webhooks to trigger actions.

**However**, I don't think additional tooling is the correct way to go about solving this issue.  #768 is already an outstanding issue for adding version checking, and there has already been some discussion on this topic.

My own opinion is we need to take another look at our current release process instead of trying to put bandaids over symptoms of a larger issue.  I've never been particularly sold on the virtues of using Homu, and while it _is_ nice to get updates on open PRs when it can no longer be merged with the base branch, it does add overhead (and another tool) that I don't think we need.  

As I mentioned in the second-last paragraph of [this comment](https://github.com/new-xkit/XKit/issues/768#issuecomment-157157923):

> Lastly, I don't think we should even be bumping version numbers in the PRs.  It's one thing for members of the Collaborators team, but it's unreasonable to expect members of the community to know how to properly increment their semver values. We've seen this in #652 where there was confusion around first converting the legacy version identifiers to semver, and then properly incrementing them.  Look at #554 and #718, where the changes become unmergeable because more recent PRs have changed the version number, and that's the _only_ conflict.  Projects like Guard [specify in their contributing guidelines](https://github.com/guard/guard/blob/master/CONTRIBUTING.md#development) that contributors should not bump the version number themselves, likely for this exact reason.

I understand that quick response and fix release times are important, but our release process is kind of a mess: we have no staging environment for testing (see #746), changes go out the door immediately after they're approved, and we're not "cutting releases" so much as every single _individual change_ results in a version increment.

I may sound a little crazy by suggesting this, but do we even _need_ version numbers for extensions at all? We don't support installing old or multiple versions of extensions, there is (currently) no compatibility checking (i.e. Y requires version `>= 1.3.6` of Z to work), and as far as I am aware, we only use versions in order to determine if an extension should be updated.  Why not do away with version numbers entirely and simply use something like the SHA of the extension's contents as a means of comparison to determine if an extension needs to be updated.  If the SHA saved to localstorage (or we could consider live-SHAing installed extensions at update time, but that would prevent people from editing extensions) doesn't match the one included in the `gallery.json`, then update the extension.

Thoughts?

**Comment by hobinjk** [comment]
@0xazure I agree that hashing is a better solution. We could even use hashes as a way of replacing the version number UI with "Latest" | "Refresh Required" | "Connection Error".

Telling contributors to not bump versions themselves is a great idea.

Thanks for the insights! I'm glad my overly complicated pipe dream led to this.

**Comment by coderpatsy** [comment]
Version numbers are handy for users to know something updated though.

**Comment by ThePsionic** [comment]
For support purposes the extensions should still have either a version or a
(partial) hash displayed in the detail screen. I assume the Week In Review
will also need to undergo a change if this goes through. @coderpatsy We can
still show the popup as usual, it just won't show the version numbers in the
detail window, just a list of updated extensions.
